web/courses.go

```diff
@@ -469,11 +479,53 @@ func (s *AutograderService) enrollTeacher(ctx context.Context, sc scm.SCM, enrol
 	})
 }
 
+// returns all enrollments for the course ID with last activity date and number of approved assignments
+func (s *AutograderService) getEnrollmentsWithActivity(courseID uint64) ([]*pb.Enrollment, error) {
+	allEnrollmentsWithSubmissions, err := s.getAllCourseSubmissions(&pb.SubmissionsForCourseRequest{CourseID: courseID, Type: pb.SubmissionsForCourseRequest_ALL})
+	if err != nil {
+		return nil, err
+	}
+	var enrollmentsWithActivity []*pb.Enrollment
+	for _, enrolLink := range allEnrollmentsWithSubmissions.Links {
+		enrol := enrolLink.Enrollment
+		var totalApproved uint64
+		for _, submissionLink := range enrolLink.Submissions {
+			if submissionLink.Submission != nil && submissionLink.Submission.Status == pb.Submission_APPROVED {
+				totalApproved++
+			}
+		}
+		enrol.TotalApproved = totalApproved
+
+		if enrol.LastActivityDate == "" {
```

Comment:
Move this test inside the loop above, and you can avoid the loop below, avoding to iterate over `enrolLink.Submissions` twice. Though, I suggest to separate the content of the loop below into a separate function, e.g. `extractLastActivity` or something.

Reply:
Fixed.